imap/ctl_cyrusdb.c

```diff
@@ -173,12 +173,69 @@ static int fixmbox(const mbentry_t *mbentry,
                mbentry->name, cyrusdb_strerror(r));
     }
 
+    /* make sure every mbentry has a uniqueid!  */
+    if (!mbentry->uniqueid) {
+        struct mailbox *mailbox = NULL;
+        struct mboxlock *namespacelock = NULL;
+        mbentry_t *copy = NULL;
+
+        r = mailbox_open_iwl(mbentry->name, &mailbox);
+        if (r) {
+            /* XXX what does it mean if there's an mbentry, but the mailbox
+             * XXX was not openable?
+             */
+            syslog(LOG_DEBUG, "%s: mailbox_open_iwl %s returned %s",
+                              __func__, mbentry->name, error_message(r));
+            goto skip_uniqueid;
+        }
+
+        if (!mailbox->uniqueid) {
+            /* yikes, no uniqueid in header either! */
+            mailbox_make_uniqueid(mailbox);
+            xsyslog(LOG_INFO, "mailbox header had no uniqueid, creating one",
+                              "mboxname=<%s> newuniqueid=<%s>",
+                              mbentry->name, mailbox->uniqueid);
```

Comment:
During the mailbox-id rollout we changed the codebase to always use the `mailbox_name()` and `mailbox_uniqueid` functions, rather than accessing these fields directly. Are these function already available in 3.4? If so, we don't need to change this part of the code when/if we pull this patch also into the main branch.

Reply:
Two things:

* those accessor functions aren't available in 3.4 anyway, but,
* even if they were, they're ambiguous with respect to whether they return the value from the _mbentry_ or from the _header_, so they're not safe when specifically trying to resolve differences between the two copies. (In practice, they currently return the _mbentry_ copy... but if that implementation detail changed, the behaviour here could break, and we wouldn't notice, because we don't experience the problems this set of fixes addresses.)

Reply:
Great, thanks for clarifying.